audiostream/src/ni/media/audio/source/gstreamer_file_source.cpp

```diff
@@ -0,0 +1,337 @@
+
+    if ( wait_for_async_operation() != GST_STATE_PLAYING )
+        throw std::runtime_error( "gstreamer_file_source: pipeline doesn't preroll into playing state" );
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+void gstreamer_file_source::fill_format_info( GstStructure*                        caps_struct,
+                                              audio::ifstream_info::container_type container )
+{
+    m_info.container( container );
+    m_info.codec( audio::ifstream_info::codec_type::mp3 );
+    m_info.lossless( false );
+
+    gint64 num_frames = 0;
+    if ( !gst_element_query_duration( m_pipeline.get(), GST_FORMAT_DEFAULT, &num_frames ) )
+        throw std::runtime_error( "gstreamer_file_source: could not query duration from gstreamer" );
+
+    m_info.num_frames( num_frames );
+
+    int sample_rate = 0;
+    if ( !gst_structure_get_int( caps_struct, "rate", &sample_rate ) )
+        throw std::runtime_error( "gstreamer_file_source: could not query sample rate from gstreamer" );
+
+    m_info.sample_rate( sample_rate );
+
+    int num_channels = 0;
+    if ( !gst_structure_get_int( caps_struct, "channels", &num_channels ) )
+        throw std::runtime_error( "gstreamer_file_source: could not query number of channels from gstreamer" );
+
+    m_info.num_channels( num_channels );
+
+    m_info.format( create_runtime_format( caps_struct ) );
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+pcm::runtime_format gstreamer_file_source::create_runtime_format( GstStructure* caps_struct )
+{
+    if ( auto format = gst_structure_get_string( caps_struct, "format" ) )
+    {
+        pcm::number_type number_type = gst_format_char_to_number_type( format[0] );
+        auto             srcDepth    = std::atol( format + 1 );
+        auto             endian      = ( strcmp( format, "BE" ) == 0 ) ? pcm::big_endian : pcm::little_endian;
+        return pcm::runtime_format( number_type, srcDepth, endian );
+    }
+    throw std::runtime_error( "gstreamer_file_source: could not get runtime format from gstreamer caps" );
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+pcm::number_type gstreamer_file_source::gst_format_char_to_number_type( const gchar format )
+{
+    if ( format == 'U' )
+        return pcm::unsigned_integer;
+    else if ( format == 'F' )
+        return pcm::floating_point;
+
+    return pcm::signed_integer;
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+void gstreamer_file_source::onPadAdded( GstElement* element, GstPad* pad, GstElement* sink )
+{
+    tGstPtr<GstPad> sinkpad( gst_element_get_static_pad( sink, "sink" ), gst_object_unref );
+    gst_pad_link( pad, sinkpad.get() );
```

Comment:
You should check for errors here, and also it can be called multiple times (container format with audio/video streams). You also probably want to ensure that you actually only link the audio stream here, and not e.g. video

Reply:
Done.